### PR TITLE
add Remainder, still need the tests

### PR DESCRIPTION
**Pull request by mrr00b00t**
Trying to add Remainder transformation, but the tests is needed

### CODE REVIEW

**Comment by mrr00b00t** [comment]
I was writing the tests but forgot that the other PR will make changes in tests/transforms.jl, so I decided not to push the tests for the Remainder for now.

File: src/transforms/remainder.jl [hunk 1]
```diff
@@ -10,18 +10,78 @@ and returns a new table with an additional column containing the
 remainder value `xₙ₊₁ = total .- (x₁ + x₂ + ⋯ + xₙ)` If the `total`
 value is not specified, then default to the maximum sum across rows.
 """
-struct Remainder <: Transform end
+struct Remainder <: Transform
+  total::Union{Nothing, Number}
```
Comment (juliohm) [inline src/transforms/remainder.jl]:
```suggestion
  total::Union{Float64,Nothing}
```

File: src/transforms/remainder.jl [hunk 2]
```diff
@@ -10,18 +10,78 @@ and returns a new table with an additional column containing the
 remainder value `xₙ₊₁ = total .- (x₁ + x₂ + ⋯ + xₙ)` If the `total`
 value is not specified, then default to the maximum sum across rows.
 """
-struct Remainder <: Transform end
+struct Remainder <: Transform
+  total::Union{Float64,Nothing}
+
+  function Remainder(total=nothing)
+    new(total)
+  end
```
Comment (juliohm) [inline src/transforms/remainder.jl]: Please remove this inner constructor and add an outer constructor with `Remainder() = Remainder(nothing)`

File: src/transforms/remainder.jl [hunk 3]
```diff
@@ -10,18 +10,78 @@ and returns a new table with an additional column containing the
...
+
+function apply(transform::Remainder, table)
+  # basic checks
+  for assertion in assertions(transform)
+    assertion(table)
+  end
+
+  # table as matrix and get the sum acros dims 2
+  X = Tables.matrix(table)
+  Σ = sum(X, dims=2)
```
Comment (juliohm) [inline src/transforms/remainder.jl]: Please use S for sum. We already use \Sigma for covariances in other parts of the code if I am not mistaken.

File: src/CoDa.jl [hunk 4]
```diff
@@ -20,7 +20,7 @@ using Printf
 import Tables
 import ScientificTypes as ST
 import TableTransforms as TT
-import TableTransforms: Transform, Stateless
+import TableTransforms: Transform, Stateless, Reject
```
Comment (juliohm) [inline src/CoDa.jl]: We are using Reject somewhere?

File: src/transforms/remainder.jl [hunk 5]
```diff
@@ -10,18 +10,76 @@ and returns a new table with an additional column containing the
...
-  # TODO
+assertions(::Type{Remainder}) = [TT.assert_continuous]
+
+function apply(transform::Remainder, table)
+  # basic checks
+  for assertion in assertions(transform)
+    assertion(table)
+  end
+
+  # table as matrix and get the sum acros dims 2
```
Comment (juliohm) [inline src/transforms/remainder.jl]:
```suggestion
  # sum for each row
```

File: src/transforms/remainder.jl [hunk 6]
```diff
@@ -10,18 +10,76 @@ and returns a new table with an additional column containing the
...
-  # TODO
+assertions(::Type{Remainder}) = [TT.assert_continuous]
+
+function apply(transform::Remainder, table)
+  # basic checks
+  for assertion in assertions(transform)
+    assertion(table)
+  end
+
+  # sum for each row
```
Comment (juliohm) [inline src/transforms/remainder.jl]:
```suggestion
  # design matrix
```

File: src/transforms/remainder.jl [hunk 7]
```diff
@@ -10,18 +10,76 @@ and returns a new table with an additional column containing the
...
+  # design matrix
+  X = Tables.matrix(table)
+  S = sum(X, dims=2)
+
+  # infer total when not specified
+  total = if !isnothing(transform.total)
+    transform.total
+  else
+    maximum(S)
+  end
```
Comment (juliohm) [inline src/transforms/remainder.jl]:
```suggestion

  # find total across rows
  total = if !isnothing(transform.total)
    transform.total
  else
    maximum(sum(X, dims=2))
  end
```

File: src/CoDa.jl (hunk 4 above)
Comment (juliohm) [reply src/CoDa.jl]: Oh I see, you're using it in the revert call. I will remove the explicit import and will use `TT.Reject` instead.

Comment (juliohm) [inline src/CoDa.jl]:
```suggestion
import TableTransforms: Transform, Stateless
```

File: src/transforms/remainder.jl [hunk 8]
```diff
@@ -10,18 +10,75 @@ and returns a new table with an additional column containing the
...
+  names = (names..., :remainder)
+  𝒯 = (; zip(names, eachcol(Z))...)
+
+  newtable = 𝒯 |> Tables.materializer(table)
+  newtable, total
 end
 
 function revert(::Remainder, newtable, cache)
-  # TODO
+  Reject(:remainder)(newtable)
```
Comment (juliohm) [inline src/transforms/remainder.jl]:
```suggestion
  TT.Reject(:remainder)(newtable)
```

File: src/transforms/remainder.jl [hunk 9]
```diff
@@ -10,18 +10,75 @@ and returns a new table with an additional column containing the
...
+  # original column names
+  names = Tables.columnnames(table)
+
+  # table as matrix and get the sum acros dims 2
+  X = Tables.matrix(table)
+  S = sum(X, dims=2)
+
+  # create a column with the remainder of each row
+  T = total .- S
+  Z = hcat(X, T)
```
Comment (juliohm) [inline src/transforms/remainder.jl]:
```suggestion
  Z = [X T]
```

File: src/transforms/remainder.jl [hunk 10]
```diff
@@ -10,18 +10,75 @@ and returns a new table with an additional column containing the
...
+  # original column names
+  names = Tables.columnnames(table)
+
+  # table as matrix and get the sum acros dims 2
+  X = Tables.matrix(table)
+  S = sum(X, dims=2)
+
+  # create a column with the remainder of each row
+  T = total .- S
+  Z = [X T]
```
Comment (juliohm) [inline src/transforms/remainder.jl]:
```suggestion
  # design matrix
  X = Tables.matrix(table)

  # create a column with the remainder
  S = sum(X, dims=2)
  Z = [X (total .- S)]
```

File: src/transforms/remainder.jl [hunk 11]
```diff
@@ -10,18 +10,74 @@ and returns a new table with an additional column containing the
...
+  else
+    maximum(sum(X, dims=2))
+  end
+
+  # original column names
+  names = Tables.columnnames(table)
+
+  # create a column with the remainder of each row
+  T = total .- S
+  Z = hcat(X, T)
```
Comment (juliohm) [inline src/transforms/remainder.jl]:
```suggestion
  # create a column with the remainder
  Z = [X (total .- S)]
```

File: src/transforms/remainder.jl [hunk 12]
```diff
@@ -10,18 +10,73 @@ and returns a new table with an additional column containing the
...
+
+  # create a column with the remainder
+  Z = [X (total .- S)]
+
+  # table with the new column
+  names = (names..., :remainder)
+  𝒯 = (; zip(names, eachcol(Z))...)
+
+  newtable = 𝒯 |> Tables.materializer(table)
+  newtable, total
```
Comment (juliohm) [inline src/transforms/remainder.jl]:
```suggestion
  𝒯 = (; zip(names, eachcol(Z))...)
  newtable = 𝒯 |> Tables.materializer(table)
  
  newtable, total
```

**Review by juliohm — CHANGES_REQUESTED** [review]
I've finished adjusting small details. Now we need two more adjustments:

1. The column :remainder should be called `total_minus_abcdef...` where `abcdef` is just a concatenation of names in the variable `names`.
2. The two function `apply` and `reapply` only differ by the definition of `total = func(transform, table, cache)`. In `apply` the function computes the total and in `reapply` the function reads the cache. Can you try to refactor the code that way so that we reduce the amount of code repetition?

File: src/transforms/remainder.jl [hunk 13]
```diff
@@ -36,47 +31,49 @@ function apply(transform::Remainder, table)
...
-  X = Tables.matrix(table)
-
   # create a column with the remainder
   S = sum(X, dims=2)
   Z = [X (total .- S)]
 
   # table with the new column
-  names = (names..., :remainder)
+  remaindercol = Symbol("total_minus_" * join(string.(names)))
+  names = (names..., remaindercol)
```
Comment (juliohm) [inline src/transforms/remainder.jl]:
```suggestion
  rname = Symbol("total_minus_" * join(string.(names)))
  names = (names..., rname)
```

File: src/transforms/remainder.jl [hunk 14]
```diff
@@ -36,47 +31,49 @@ function apply(transform::Remainder, table)
...
   newtable, total
 end
+
+function apply(transform::Remainder, table)
+  _apply(transform, table, _cache(transform, table))
+end
+
+function revert(::Remainder, newtable, cache)
+  names = Tables.columnnames(newtable)
+  TT.Reject(names[end])(newtable)
```
Comment (juliohm) [inline src/transforms/remainder.jl]:
```suggestion
  TT.Reject(last(names))(newtable)
```

**Review by juliohm — APPROVED** [review]

File: src/transforms/remainder.jl [hunk 15]
```diff
@@ -10,18 +10,70 @@ and returns a new table with an additional column containing the
...
+  X = Tables.matrix(table)
+
+  # find total across rows
+  total = if !isnothing(transform.total)
+    transform.total
+  else
+    maximum(sum(X, dims=2))
+  end
+
+  total
```
Comment (juliohm) [inline src/transforms/remainder.jl]:
```suggestion
  if !isnothing(transform.total)
    transform.total
  else
    maximum(sum(X, dims=2))
  end
```

File: src/transforms/remainder.jl [hunk 16]
```diff
@@ -10,18 +10,68 @@ and returns a new table with an additional column containing the
...
+  names = (names..., rname)
+  𝒯 = (; zip(names, eachcol(Z))...)
+  newtable = 𝒯 |> Tables.materializer(table)
+  
+  newtable, total
+end
+
+function apply(transform::Remainder, table)
+  _apply(transform, table, _cache(transform, table))
 end
```
Comment (juliohm) [inline src/transforms/remainder.jl]:
```suggestion
apply(transform::Remainder, table) = _apply(transform, table, _cache(transform, table))
```

File: src/transforms/remainder.jl [hunk 17]
```diff
@@ -10,18 +10,66 @@ and returns a new table with an additional column containing the
...
-  # TODO
+  names = Tables.columnnames(newtable)
+  TT.Reject(last(names))(newtable)
 end
 
-function reapply(::Remainder, table, cache)
-  # TODO
+function reapply(transform::Remainder, table, cache)
+  _apply(transform, table, cache)
 end
```
Comment (juliohm) [inline src/transforms/remainder.jl]:
```suggestion
reapply(transform::Remainder, table, cache) = _apply(transform, table, cache)
```

**Comment by juliohm** [comment]
Closing in favor of #28

**Pull request closed without merging**